# Déjame escuchar música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>

Como analista de datos, debemos analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos comprobar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de `Springfield` y `Shelbyville`. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:

Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.


### Etapas

Los datos del comportamiento del usuario se almacenan en un archivo CSV `music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que debemos examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:

 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.

[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [4]:
# Leer el archivo y almacenarlo en df
#df = pd.read_csv('music_project_en.csv')
df= pd.read_csv("/datasets/music_project_en.csv")

Muestra las 10 primeras filas de la tabla:

In [5]:
# Obtener las 10 primeras filas de la tabla df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [6]:
# Obtener la información general sobre nuestros datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `'userID'`: identificador del usuario o la usuaria;

- `'Track'`: título de la canción;

- `'artist'`: nombre del artista;

- `'genre'`: género de la pista;

- `'City'`: ciudad del usuario o la usuaria;

- `'time'`: la hora exacta en la que se reprodujo la canción;

- `'Day'`: día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:

1. Algunos encabezados están en mayúsculas, otros en minúsculas.

2. Hay espacios en algunos encabezados.

3. Uso de las convenciones de nomenclatura, respetar el estilo `snake_case`. 

### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

* Todos los datos en las columnas parecen ser del tipo object, lo que sugiere que contienen cadenas de texto.

* Para comprender mejor el contenido de las columnas, podemos examinar algunas filas utilizando métodos como `.sample`, `head()` o `tail()` para obtener una muestra de la base de datos.

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

* Antes de confirmar si hay suficientes datos para responder a la hipótesis, necesitamos comprender mejor la naturaleza de la hipótesis y los tipos de análisis que requerirá.

* Podríamos explorar la distribución de los datos en función de las ciudades, los días de la semana y las horas del día para determinar si hay suficientes observaciones para realizar comparaciones significativas entre los grupos.

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

* Podríamos necesitar estandarizar los nombres de las columnas para mejorar la coherencia y facilitar el análisis.

* Es importante verificar si hay valores ausentes, duplicados o tipos de datos incorrectos en el conjunto de datos.

* Podemos utilizar métodos como `isnull()`, `duplicated()`, `info()`, `describe()` para identificar y abordar cualquier problema potencial en los datos antes de proceder con el análisis.

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corregiremos el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [7]:
# Muestra los nombres de las columnas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.


Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [8]:
# Crear lista vacía para las nuevas columnas
new_col_names = []

# Bucle en los encabezados poniendo todo en minúsculas
for column in df.columns:
    name_lower = column.lower()

    # Agrega el nuevo nombre a la lista de nuevos nombres de columna
    new_col_names.append(name_lower)

# Muestra los encabezados de la tabla actualizados
print(new_col_names)

['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day']


Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [9]:
# Crea lista vacía para los nuevos nombres de columna
new_col_names = []

# Bucle en los encabezados poniendo todo en minúsculas y eliminando espacios al principio y al final
for column in df.columns:
    name_stripped = column.strip().lower()

    # Agrega el nuevo nombre a la lista de nuevos nombres de columna
    new_col_names.append(name_stripped)

# Renombra las columnas en el DataFrame utilizando los nuevos nombres
df.columns = new_col_names

# Muestra los encabezados de la tabla actualizados
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [10]:
# Cambia el nombre de la columna 'userid' a 'user_id'
df.rename(columns={'userid': 'user_id'}, inplace=True)

# Muestra los nombres de todas las columnas después del cambio
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


Comprueba el resultado. Muestra los encabezados una vez más:

In [11]:
# Muestra los nombres de todas las columnas después del cambio
df.columns

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [12]:
# Calcular el número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [13]:
# Lista de columnas que necesitan reemplazo de valores ausentes
columns_to_replace = ['track', 'artist', 'genre']

# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'
for column in columns_to_replace:
    df[column].fillna('unknown', inplace=True)

# Muestra los primeros registros del DataFrame para verificar el reemplazo
df.head(10)

user_id                        track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist           unknown   dance   

          city      time        day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [14]:
# Contar valores ausentes
print(df.isna().sum())

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [15]:
# Contar duplicados explícitos
df.duplicated().sum()

3826

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [16]:
# Eliminar duplicados explícitos
df.drop_duplicates(inplace=True)

Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [17]:
# Comprobar de nuevo si hay duplicados
df.duplicated().sum()

0

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [18]:
# Extraer la columna 'genre' del DataFrame y obtener valores únicos ordenados alfabeticamente
unique_genres = df['genre'].unique()
unique_genres_sorted = sorted(unique_genres)

print(unique_genres_sorted)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`

* `hop`

* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:

* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.

* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [19]:
# Función para reemplazar duplicados implícitos
def replace_wrong_genres(wrong_genres, correct_genre):
    # Itera sobre la lista de géneros incorrectos
    for wrong_genre in wrong_genres:
        # Reemplaza los valores incorrectos por el valor correcto en la columna 'genre'
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

# Lista de géneros incorrectos y el género correcto
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [20]:
# Llama a la función para realizar los reemplazos
replace_wrong_genres(wrong_genres, correct_genre)

Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [21]:
# Comprobación de duplicados implícitos
print(sorted(df['genre'].unique()))

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

Al analizar los duplicados en la tabla, observé dos tipos: duplicados **explícitos** y duplicados **implícitos**. Los duplicados explícitos se refieren a filas completas que son idénticas en todos sus valores. Utilicé el método **duplicated()** seguido de **sum()** para contar estos duplicados y luego utilicé **drop_duplicates()** para eliminarlos.

Para abordar los duplicados implícitos en la columna **'genre'**, primero extraje los valores únicos de esta columna y los ordené alfabéticamente. Luego, identifiqué duplicados implícitos del género **'hiphop'**, como **'hip', 'hop', 'hip-hop'**.

Para eliminar estos duplicados implícitos, creé una función llamada **replace_wrong_genres()** que toma una lista de géneros incorrectos y un género correcto como parámetros. Dentro de esta función, utilicé un bucle **for** para iterar sobre la lista de géneros incorrectos y reemplacé cada uno de ellos por el género correcto utilizando el método **replace()** en la columna **'genre'** del DataFrame.

[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [22]:
# Agrupar los datos por ciudad y sumar la cantidad de canciones reproducidas en cada grupo
canciones_reproducidas_por_ciudad = df.groupby('city')['track'].count()

# Mostrar el resultado
print(canciones_reproducidas_por_ciudad)

city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


**Springfield** muestra una actividad de reproducción considerablemente **mayor**: La cantidad total de canciones reproducidas en **Springfield** (**42741**) es mucho mayor que en **Shelbyville** (**18512**). Esto sugiere que los usuarios y las usuarias de Springfield están más activos en la reproducción de música en comparación con los de Shelbyville.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [23]:
# Filtrar los datos para incluir solo los días de interés
df_filtered = df[(df['day'] == 'Monday') | (df['day'] == 'Wednesday') | (df['day'] == 'Friday')]

# Agrupar los datos por día de la semana y contar la cantidad de canciones reproducidas en cada grupo
canciones_por_dia = df_filtered.groupby('day')['track'].count()

# Mostrar el resultado
print(canciones_por_dia)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


Los datos revelan que los usuarios y las usuarias muestran patrones distintos en la reproducción de música a lo largo de la semana. El **viernes** se destaca como el día con la **mayor actividad**, con un total de **21,840 canciones reproducidas**, posiblemente debido al **inicio del fin de semana**. En contraste, **lunes y miércoles** muestran cantidades **similares** de reproducciones, sugiriendo una **estabilidad en la actividad de reproducción durante la semana laboral**. Estos hallazgos apuntan a una **variación en el comportamiento de consumo de música según el día de la semana**, destacando la influencia de factores como el calendario laboral y las rutinas diarias en la actividad de reproducción de música.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [24]:
def number_tracks(day, city):
    # Filtrar por día
    filtered_by_day = df[df['day'] == day]
    
    # Filtrar por ciudad
    filtered_by_city = filtered_by_day[filtered_by_day['city'] == city]
    
    # Contar el número de canciones reproducidas
    tracks_count = filtered_by_city['track'].count()
    
    # Devolver el número de canciones reproducidas
    return tracks_count

Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [25]:
# El número de canciones reproducidas en Springfield el lunes
springfield_mon = number_tracks('Monday', 'Springfield')

print(springfield_mon)

15740


In [26]:
# El número de canciones reproducidas en Shelbyville el lunes
shelbyville_mon = number_tracks('Monday', 'Shelbyville')

print(shelbyville_mon)

5614


In [27]:
# El número de canciones reproducidas en Springfield el miércoles
springfield_wed = number_tracks('Wednesday', 'Springfield')

print(springfield_wed)

11056


In [28]:
# El número de canciones reproducidas en Shelbyville el miércoles
shelbyville_wed = number_tracks('Wednesday', 'Shelbyville')

print(shelbyville_wed)

7003


In [29]:
# El número de canciones reproducidas en Springfield el viernes
springfield_fri = number_tracks('Friday', 'Springfield')

print(springfield_fri)

15945


In [30]:
# El número de canciones reproducidas en Shelbyville el viernes
shelbyville_fri = number_tracks('Friday', 'Shelbyville')

print(shelbyville_fri)

5895


**Conclusiones**

Se confirma que existen diferencias notables en la cantidad de canciones reproducidas entre las ciudades de Springfield y Shelbyville. Se destaca una clara tendencia, donde el viernes emerge como el día con mayor actividad de reproducción considerando ambas ciudades. Es interesante notar que, si bien el viernes es el día predominante en Springfield, el miércoles es el día con mayor actividad en Shelbyville. 

[Volver a Contenidos](#back)

# Conclusiones <a id='end'></a>

Basádo en los resultados, se puede concluir que la hipótesis es válida. Los datos muestran claramente diferencias tanto en la actividad de reproducción de música entre las ciudades de Springfield y Shelbyville, como en la actividad según el día de la semana. En Springfield, se reproduce una cantidad considerablemente mayor de canciones en comparación con Shelbyville, lo que sugiere una mayor actividad de reproducción de música en esta ciudad. Además, se observa que los viernes son los días con mayor actividad de reproducción de música en Springfield, en cambio para Shelbylle es el miércoles. Estas variaciones indican que tanto la ubicación geográfica como el día de la semana influyen en el comportamiento de reproducción de música de los usuarios/as.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

[Volver a Contenidos](#back)